### PR TITLE
Add PR template (SQWG trial).

### PR DESCRIPTION
**Pull request by gavanderhoorn**
Similar to https://github.com/ament/ament_lint/pull/129.

I've adapted the text for this repository (and ROS 2).

This is part of the trial started by the members of the software quality working group trying to get more people involved in maintainership and increase community participation in these processes.

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
@itualami: fyi.

**Comment by thomas-moulard** [comment]
FYI here is something which has been stuck in my backburner for the last couple of weeks:

```
### Your checklist for this pull request
🚨 Please review the [guidelines for contributing]( https://index.ros.org/doc/ros2/Contributing/Developer-Guide/) to this repository.

- [ ] Make sure that your commit messages include a [DCO]( https://discourse.ros.org/t/starting-to-enforce-developer-certificate-of-origin-dco-for-some-ros-2-repos/7420).
- [ ] Make sure your commit messages follow [good practices]( https://chris.beams.io/posts/git-commit/).
- [ ] Make sure your PR is minimal. Unrelated changes should be submitted separately.

### Description

Please describe your pull request.

### Testing

PRs must be tested on each Tier 1 platform as documented by the [REP-2000]( http://www.ros.org/reps/rep-2000.html).

Tier 1:
- [ ] Linux Ubuntu x86-64
- [ ] Linux Ubuntu arch64
- [ ] OS X Testing
- [ ] Windows Testing

Other tested Tier 2 and Tier 3 platforms:

### Release Engineering

- Should this PR cherry-picked to the next patch release? Yes/No (choose No if unsure, only possible for PRs preserving ABI compatibility)
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<!--
+Thanks for submitting a Pull Request!
+
+Please shortly explain your contribution, and if fixing an issue from the tracker, please add a link to the issue.
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]: Instead of "add a link", it would be better to use the GitHub PR->issue automatic tracking.  Something like:

`if fixing an issue from the tracker, please add "Fixes #XXX", replacing XXX with the issue number`

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+-->
+
+### Description
+
+Add your description here
+
+
+### Checklist
+
+- [ ] Did you change how any of these packages work? Added new functionality? Make sure to update the documentation
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]: To keep the "voice" consistent, I'd go with something more along the lines of:

```
- [ ] If you have changed how any of the APIs worked, or added new APIs, please update the API documentation.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+- [ ] If you added a new package: make sure to add documentation for it
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]: I'd be a bit more specific on where to add documentation, something like:


Additionally, I'd add another bullet along the lines of:

```
- [ ] If you've added a new feature or new package, make sure to update the documentation at https://index.ros.org/doc/ros2/ describing the new feature.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+- [ ] If needed, please include (updated) tests with your PR
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]: Here I would say something more along the lines of:

```
- [ ] Add or update tests corresponding to your PR.  If tests cannot be added for some reason, be sure to explain why in the description.
```

**Comment by clalancette** [comment]
@gavanderhoorn Friendly ping on this.

**Comment by Ridhwanluthra** [comment]
@clalancette I had volunteered to incorporate the changes requested, I am sorry for the delay on my part.
I am only changing on this repo and once we finalise on a template I would update on the others accordingly.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<!--
+Thanks for submitting a Pull Request!
+
+Please shortly explain your contribution, and if fixing an issue from the tracker, please add "Fixes #XXX", replacing XXX with the issue number.
+
+Be ensure that your contribution follows the [Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/).
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
Be sure that your contribution follows the [Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/).
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+Be sure to go over each item in the list below before submitting your pull request.
+-->
+
+### Description
+
+Add your description here
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
Add your description here
```

Briefly describe what your changes are doing.  It is more important to mention why the changes are necessary, rather than the how.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+### Checklist
+
+- [ ] Make sure that your commit messages include a [DCO]( https://discourse.ros.org/t/starting-to-enforce-developer-certificate-of-origin-dco-for-some-ros-2-repos/7420).
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] Make sure that your commit messages include a [DCO](https://discourse.ros.org/t/starting-to-enforce-developer-certificate-of-origin-dco-for-some-ros-2-repos/7420).
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+- [ ] Make sure your commit messages follow [good practices]( https://chris.beams.io/posts/git-commit/).
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] Make sure your commit messages follow [good practices](https://chris.beams.io/posts/git-commit/).
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+PRs must be tested on each Tier 1 platform as documented by the [REP-2000]( http://www.ros.org/reps/rep-2000.html).
+
+Tier 1:
+
+- [ ] Linux Ubuntu x86-64
+- [ ] Linux Ubuntu arch64
+- [ ] OS X Testing
+- [ ] Windows Testing
+
+Other tested Tier 2 and Tier 3 platforms:
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]: Just remove this, since it is an empty section.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+### Release Engineering
+
+- Should this PR cherry-picked to the next patch release? Yes/No (choose No if unsure, only possible for PRs preserving ABI compatibility)
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] Should this PR be backported to supported ROS 2 releases?  Changes that break API or ABI are not generally eligible to be backported.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+Add your description here
+
+### Checklist
+
+- [ ] Make sure that your commit messages include a [DCO]( https://discourse.ros.org/t/starting-to-enforce-developer-certificate-of-origin-dco-for-some-ros-2-repos/7420).
+- [ ] Make sure your commit messages follow [good practices]( https://chris.beams.io/posts/git-commit/).
+- [ ] Make sure your PR is minimal. Unrelated changes should be submitted separately.
+- [ ] If you have changed how any of the APIs worked, or added new APIs, please update the API documentation.
+- [ ] If you've added a new feature or new package, make sure to update the documentation at https://index.ros.org/doc/ros2/ describing the new feature.
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] If you've added a new feature or new package, make sure to update the [documentation](https://index.ros.org/doc/ros2/),  describing the new feature.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+- [ ] Add or update tests corresponding to your PR.  If tests cannot be added for some reason, be sure to explain why in the description.
+- [ ] While waiting for someone to review your PR, please consider reviewing [another open pull request](https://github.com/ros2/rcl/pulls) to support the maintainers of RCL. Refer to the [ROS code review guide](https://github.com/rosin-project/ros_code_review_guide/blob/master/README.md) for general reviewing guidelines and [Quality Guide: Ensuring code quality](https://index.ros.org/doc/ros2/Contributing/Quality-Guide/) for ROS 2 specific guidelines
+
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]: As @dirk-thomas pointed out earlier, we should also have a bullet (early in the list) about passing tests locally.

**Comment by Ridhwanluthra** [comment]
I was thinking that maybe we should change the wording for the lines in the checklist since its a checklist.
for eg, instead of having
`[ ] Make sure that your commit messages include a DCO.`
We have 
`[ ] The commit message includes a DCO.`

**Comment by clalancette** [comment]
> I was thinking that maybe we should change the wording for the lines in the checklist since its a checklist.
> for eg, instead of having
> `[ ] Make sure that your commit messages include a DCO.`
> We have
> `[ ] The commit message includes a DCO.`

Yeah, I like that.  It flows better.

**Comment by Ridhwanluthra** [comment]
Alright, I will go ahead and make those changes

**Comment by Ridhwanluthra** [comment]
I am not sure how to word the PR review checklist item, I have asked some others for help. The others are all updated

**Comment by itualami** [comment]
> I am not sure how to word the PR review checklist item, I have asked some others for help. The others are all updated

The language seems fine to me.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+### Description
+
+Briefly describe what your changes are doing. It is more important to mention why the changes are necessary, rather than the how.
+
+### Checklist
+
+- [ ] The commit messages include a [DCO](https://discourse.ros.org/t/starting-to-enforce-developer-certificate-of-origin-dco-for-some-ros-2-repos/7420).
+- [ ] The commit messages follow [good practices](https://chris.beams.io/posts/git-commit/).
+- [ ] The build passes tests locally.
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] The changes passes tests locally.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+- [ ] The PR is minimal and any unrelated changes are submitted separately.
+- [ ] The API documentation has been updated based on changes in APIs or addition of new APIs, if any.
+- [ ] The [documentation](https://index.ros.org/doc/ros2/) has been updated to describe the new feature or package added, if any.
+- [ ] Added or updated tests corresponding to the PR. If tests cannot be added for some reason, it has been explained in the description.
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] Tests have been added or updated to cover the changes. If tests cannot be added for some reason, it has been explained in the description.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+- [ ] While waiting for someone to review your PR, please consider reviewing [another open pull request](https://github.com/ros2/rcl/pulls) to support the maintainers of RCL. Refer to the [ROS code review guide](https://github.com/rosin-project/ros_code_review_guide/blob/master/README.md) for general reviewing guidelines and [Quality Guide: Ensuring code quality](https://index.ros.org/doc/ros2/Contributing/Quality-Guide/) for ROS 2 specific guidelines
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]: I'd think we should probably take this out of the checklist and just make it a comment after the block.  It's not really an actionable thing on *this* PR.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+### Testing
+
+PRs must be tested on each Tier 1 platform as documented by the [REP-2000]( http://www.ros.org/reps/rep-2000.html).
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
PRs must be tested on each Tier 1 platform as documented by the [REP-2000](http://www.ros.org/reps/rep-2000.html).
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+
+- [ ] Linux Ubuntu x86-64
+- [ ] Linux Ubuntu arch64
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] Linux Ubuntu arm64
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+- [ ] The API documentation has been updated based on changes in APIs or addition of new APIs, if any.
+- [ ] The [documentation](https://index.ros.org/doc/ros2/) has been updated to describe the new feature or package added, if any.
+- [ ] Added or updated tests corresponding to the PR. If tests cannot be added for some reason, it has been explained in the description.
+- [ ] While waiting for someone to review your PR, please consider reviewing [another open pull request](https://github.com/ros2/rcl/pulls) to support the maintainers of RCL. Refer to the [ROS code review guide](https://github.com/rosin-project/ros_code_review_guide/blob/master/README.md) for general reviewing guidelines and [Quality Guide: Ensuring code quality](https://index.ros.org/doc/ros2/Contributing/Quality-Guide/) for ROS 2 specific guidelines
+
+### Testing
+
+PRs must be tested on each Tier 1 platform as documented by the [REP-2000]( http://www.ros.org/reps/rep-2000.html).
+
+- [ ] Linux Ubuntu x86-64
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] Linux Ubuntu amd64
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+- [ ] Linux Ubuntu arch64
+- [ ] OS X Testing
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] MacOS Sierra
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+- [ ] Windows Testing
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] Windows 10 (VS2019)
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+### Release Engineering
+
+- [ ] This PR shoub be backported to supported ROS 2 releases. I understand changes that break API or ABI are not generally eligible to be backported.
```
Comment (clalancette) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] This PR should be backported to supported ROS 2 releases. I understand changes that break API or ABI are not generally eligible to be backported.
```

**Comment by mabelzhang** [comment]
@Ridhwanluthra Could you address the comments in the review above? Thank you.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 15 above)
Comment (clalancette) [reply .github/PULL_REQUEST_TEMPLATE.md]: This comment is still open.

Comment (Ridhwanluthra) [reply .github/PULL_REQUEST_TEMPLATE.md]: Yes, I was just discussing this with the sqwg group once.

**Comment by Ridhwanluthra** [comment]
@clalancette Based on discussions with the SQWG group we have changed the request for PR review to ensure better metrics generation which would help us with further initiatives.

**Comment by clalancette** [comment]
The PR testing failed because this PR is targeting master, but we haven't released all of the other stuff from Eloquent necessary yet.  So we can ignore that (especially since this doesn't touch any code).

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+### Checklist
+
+- [ ] The commit messages include a [DCO](https://discourse.ros.org/t/starting-to-enforce-developer-certificate-of-origin-dco-for-some-ros-2-repos/7420).
+- [ ] The commit messages follow [good practices](https://chris.beams.io/posts/git-commit/).
+- [ ] Tests have been added or updated to cover the changes. If tests cannot be added for some reason, it has been explained in the description.
+- [ ] The changes passes tests locally (see [here](https://index.ros.org/doc/ros2/Tutorials/Colcon-Tutorial) for how to run tests.
+- [ ] The PR is minimal and any unrelated changes are submitted separately.
+- [ ] The API documentation has been updated based on changes in APIs or addition of new APIs, if any.
+- [ ] The [documentation](https://index.ros.org/doc/ros2/) has been updated to describe the new feature or package added, if any.
```
Comment (dirk-thomas) [inline .github/PULL_REQUEST_TEMPLATE.md]: For many kind of changes https://index.ros.org/doc/ros2/ isn't relevant. So this might confuse a contributor.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+
+### Testing
+
+PRs must be tested on each Tier 1 platform as documented by the [REP-2000](http://www.ros.org/reps/rep-2000.html).
+
+- [ ] Linux Ubuntu amd64
+- [ ] Linux Ubuntu arm64
+- [ ] MacOS Sierra
+- [ ] Windows 10 (VS2019)
```
Comment (dirk-thomas) [inline .github/PULL_REQUEST_TEMPLATE.md]: Since a contributor creating a pull request doesn't have the ability to influence what platforms the CI systems checks I am not sure what the purpose of this list would be?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+- [ ] The changes passes tests locally (see [here](https://index.ros.org/doc/ros2/Tutorials/Colcon-Tutorial) for how to run tests.
+- [ ] The PR is minimal and any unrelated changes are submitted separately.
+- [ ] The API documentation has been updated based on changes in APIs or addition of new APIs, if any.
+- [ ] The [documentation](https://index.ros.org/doc/ros2/) has been updated to describe the new feature or package added, if any.
+
+- Reviewing pull requests takes time and effort. While we are reviewing your pull request, please consider supporting the community by reviewing another open pull request. Refer to the [ROS code review guide](https://github.com/rosin-project/ros_code_review_guide/blob/master/README.md) for general reviewing guidelines and [Quality Guide: Ensuring code quality](https://index.ros.org/doc/ros2/Contributing/Quality-Guide/) for ROS 2 specific guidelines.
+  - [ ] I will review (or have reviewed) another person’s PR. PR tracker No.: ___________________
+  - [ ] I’m willing to review another person’s PR and would like a suggestion.
+  - [ ] I do not want to review another person’s PR.
```
Comment (mjcarroll) [inline .github/PULL_REQUEST_TEMPLATE.md]: Is this something that could be put in a comment rather than in the PR body?  I appreciate the idea of getting other people to contribute to reviewing PRs, but I would be concerned that this would be a lot of noise when going through many PRs.

Comment (itualami) [reply .github/PULL_REQUEST_TEMPLATE.md]: > Is this something that could be put in a comment rather than in the PR body? I appreciate the idea of getting other people to contribute to reviewing PRs, but I would be concerned that this would be a lot of noise when going through many PRs.

We discussed this in the ROS Quality Working group and we thought that static comments tend to be ignored by developers. Hence, we suggested something actionable to incite the contributor to participate in reviewing other PRs.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 23 above)
Comment (Ridhwanluthra) [reply .github/PULL_REQUEST_TEMPLATE.md]: @thomas-moulard Could you please comment on this?

**Comment by Ridhwanluthra** [comment]
> Thanks for all of the back and forth. I made a couple of minor edits on top of yours to rearrange the items and to add a link on how to run tests with colcon. I'm now happy with this, so I'll approve it; please take another look to make sure you agree with my edits, and then we can merge this.

From our end, this seems to be appropriate. Once it is merge I will go ahead and update the other repositories.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 22 above)
Comment (Ridhwanluthra) [reply .github/PULL_REQUEST_TEMPLATE.md]: Should we eliminate this or provide something else along with it?

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 24 above)
Comment (max-krichenbauer) [reply .github/PULL_REQUEST_TEMPLATE.md]: It should be mentioned that a similar pattern is already used by other repositories, most notably [ros-planning/moveit](https://github.com/ros-planning/moveit/blob/master/.github/PULL_REQUEST_TEMPLATE.md).

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 23 above)
Comment (thomas-moulard) [reply .github/PULL_REQUEST_TEMPLATE.md]: Maybe that should be changed. I would add a comment to say something that "Only contributions passing CI on all Tier 1 platform will be merged".
Some people may not realize that ROS 2 needs to compiled on Windows and OS X too, setting this expectation upfront could be helpful?

Comment (Ridhwanluthra) [reply .github/PULL_REQUEST_TEMPLATE.md]: @thomas-moulard @dirk-thomas Please check if the new wordings seem alright?

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/next-ros-quality-assurance-working-group-february-2020/12571/1

**Comment by kscottz** [comment]
I was asked by the QAWG to see if I could get this merged. This is a simple enough request that we should be able to iterate on it fairly quickly instead of trying to solve all the problems in a single PR that sits around **_for a year_**.  @gavanderhoorn it appears all that needs to happen is for you to sign your DCO and for someone else to merge it. Could we get this merged by the end of the week?

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/next-ros-quality-assurance-working-group-march-2020/12914/1

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/april-quality-assurance-working-group-meeting/13402/1

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/may-quality-assurance-working-group-meeting/13812/1

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/may-quality-assurance-working-group-meeting/13812/6
